WorkloadManagementSystem/DB/PilotsLoggingDB.py

```diff
@@ -55,33 +49,58 @@ def __init__( self ):
     self.dbPass = dbParameters['Password']
     self.dbName = dbParameters['DBName']
 
-    self.__initializeConnection( 'WorkloadManagement/PilotsLoggingDB' )
+    #These are the list of tables that will be created.
+    #They can be extended in an extension module
+    self.tablesList = getattr(Utils.voimport( 'DIRAC.WorkloadManagementSystem.DB.PilotsLoggingDB' ),
+                              'TABLESLIST')
+
+    self.__initializeConnection()
     resp = self.__initializeDB( )
     if not resp['OK']:
       raise Exception( "Couldn't create tables: " + resp['Message'] )
 
   ##########################################################################################
 
-  def __initializeConnection( self, dbPath ):
-
-    self.engine = create_engine( 'mysql://%s:%s@%s:%s/%s'
-                                 % (self.dbUser, self.dbPass, self.dbHost, self.dbPort, self.dbName),
-                                 pool_recycle = 3600, echo_pool = True )
+  def __initializeConnection( self ):
+    """
+    This should be in a base class eventually
+    """
+
+    self.engine = create_engine( 'mysql://%s:%s@%s:%s/%s' % (self.dbUser,
+                                                             self.dbPass,
+                                                             self.dbHost,
+                                                             self.dbPort,
+                                                             self.dbName),
+                                 pool_recycle = 3600,
+                                 echo_pool = True,
+                                 echo = self.log.getLevel() == 'DEBUG' )
     self.sqlalchemySession = scoped_session( sessionmaker( bind = self.engine ) )
     self.inspector = Inspector.from_engine( self.engine )
 
   ##########################################################################################
   def __initializeDB( self ):
+    """
+    Create the tables, if they are not there yet
+    """
 
     tablesInDB = self.inspector.get_table_names( )
 
-    if not 'PilotsLogging' in tablesInDB:
-      try:
-        PilotsLogging.__table__.create( self.engine )
-      except SQLAlchemyError as e:
-        return S_ERROR( DErrno.ESQLA, e )
-    else:
-      gLogger.debug( "Table PilotsLogging exists" )
+    for table in self.tablesList:
+      if not table in tablesInDB:
+        found = False
+        #is it in the extension? (fully or extended)
+        for ext in gConfig.getValue( 'DIRAC/Extensions', [] ):
+          try:
+            getattr(__import__(ext + __name__, globals(), locals(), [table]), table).__table__.create( self.engine ) #pylint: disable=no-member
+            found = True
+            break
+          except (ImportError, AttributeError):
+            continue
+          # If not found in extensions, import it from DIRAC base.
+          if not found:
```

Comment:
The if statement needs to go outside the for loop. Otherwise it is never executed, because either line 94 gives an exception, in which case the except will continue and skip `if not found`, or line 94 will not throw an exception and found will be true and the loop will be left.

Reply:
Indeed! Good catch!